numpyro/hmc_util.py

```diff
@@ -268,3 +271,321 @@ def update_fn(t, accept_prob, z_flat, state):
         return state
 
     return init_fn, update_fn
+
+
+# TODO: we don't need to store z_proposal_pe, z_proposal_grad if we don't need
+# to cache pe and grad for the next update step
+_TreeInfo = namedtuple("_TreeInfo", ["z_left", "r_left", "z_left_grad",
+                                     "z_right", "r_right", "z_right_grad",
+                                     "z_proposal", "z_proposal_pe", "z_proposal_grad",
+                                     "depth", "weight", "r_sum", "turning", "diverging",
+                                     "sum_accept_probs", "num_proposals"])
+
+
+# let JAX recognize _TreeInfo structure
+# ref: https://github.com/google/jax/issues/446
+# TODO: remove this when namedtuple is supported in JAX
+register_pytree_node(
+    _TreeInfo,
+    lambda xs: (tuple(xs), None),
+    lambda _, xs: _TreeInfo(*xs)
+)
+
+
+def _is_turning(inverse_mass_matrix, r_left, r_right, r_sum):
+    r_left, _ = ravel_pytree(r_left)
+    r_right, _ = ravel_pytree(r_right)
+    r_sum, _ = ravel_pytree(r_sum)
+
+    if inverse_mass_matrix.ndim == 2:
+        v_left = np.matmul(inverse_mass_matrix, r_left)
+        v_right = np.matmul(inverse_mass_matrix, r_right)
+    elif inverse_mass_matrix.ndim == 1:
+        v_left = np.multiply(inverse_mass_matrix, r_left)
+        v_right = np.multiply(inverse_mass_matrix, r_right)
+
+    # This implements dynamic termination criterion (ref [2], section A.4.2).
+    turning_at_left = np.dot(v_left, r_sum - r_left) <= 0
+    turning_at_right = np.dot(v_right, r_sum - r_right) <= 0
+    return turning_at_left | turning_at_right
+
+
+def _uniform_transition_prob(current_subtree, new_subtree, rng):
```

Comment:
nit: rename `prob` to `kernel` to match the notion is NUTS paper, rename `subtree -> tree` too, and drop `rng`.